# Cross Validation

This code can be used as a template for KFold and Monte Carlo Cross Validation.

In [15]:
import numpy as np
import pandas as pd

In [16]:
data = pd.read_csv("diagnosis.csv") #You can change data adress from here


In [17]:
Y=data.iloc[:,6:7]#Y = Target data
X=data.drop(columns=['w'])



In [18]:
from sklearn.naive_bayes import GaussianNB #You can change 
gNB=GaussianNB()                           #model from here

# KFold with cross_val_score

In [19]:
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import KFold
kf=KFold(n_splits=5)

In [20]:
print("Cross validation scores with KFold")
cross_val_score(gNB,X,Y.values.reshape(-1,),cv=kf)
print("Cross validation scores with KFold mean.")
cross_val_score(gNB,X,Y.values.reshape(-1,),cv=kf).mean()

Cross validation scores with KFold
Cross validation scores with KFold mean.


0.825

# KFold with code


Let's write an alternative code that does the same function as the cross_val_score function above.

In [21]:
def get_score(model, X_train,X_test,Y_train,Y_test):
    model.fit(X_train,Y_train)
    return model.score(X_test,Y_test)

In [22]:
gNB_list=[]
for train_index,test_index in kf.split(X):
    X_train,X_test,Y_train,Y_test=X.iloc[train_index],X.iloc[test_index],\
                                  Y.iloc[train_index],Y.iloc[test_index]
    
    gNB_list.append(get_score(gNB,X_train,X_test,Y_train.values.reshape(-1,),Y_test.values.reshape(-1,)))
        

In [23]:
print("Cross validation scores with KFold")
gNB_list


Cross validation scores with KFold


[1.0, 1.0, 0.5833333333333334, 0.9166666666666666, 0.625]

# Monte Carlo Cross Validation (Shuffle Split)

MC is referred to as sf in the sklearn library.

In [24]:
from  sklearn.model_selection import ShuffleSplit 

In [25]:
mc=ShuffleSplit(n_splits=5,test_size=0.33,random_state=7)

In [26]:
print("Cross validation scores with Monte Carlo Cross Validation")
cross_val_score(gNB,X,Y.values.reshape(-1,),cv=mc).mean()

Cross validation scores with Monte Carlo Cross Validation


0.7849999999999999

When we compare Monte Carlo and KFold, we see that KFold is a more suitable model for our data set with a score of 0.825.